code/GatewayFieldsFactory.php

```diff
@@ -320,47 +340,118 @@ protected function cullForGateway(&$fields, $defaults = array())
+     */
+    public function getGlobalFieldName($defaultName)
+    {
+        $renameMap = $this->getRenameConfig();
+
+        if (is_array($renameMap) && array_key_exists($defaultName, $renameMap)) {
+            return $renameMap[ $defaultName ];
         }
 
-        $stack = array(array_flip($globalMap));
-        
-        if (array_key_exists($this->gateway, $renameMap) && !empty($renameMap[$this->gateway])) {
-            $stack[] = $gatewayMap = array_flip($renameMap[$this->gateway]);
+        return false;
+    }
+
+    /**
+     * Fetches the prefix for a gateway, or returns false
+     *
+     * @param null $gateway
+     *
+     * @return bool|string Returns false if no gateway prefix found, otherwise returns the prefix
+     */
+    public function getGatewayPrefix($gateway = null)
+    {
+        return $this->getGatewayFieldName('prefix', $gateway);
+    }
+
+    /**
+     * Fetches custom name for a gateway field from the rename map, or returns false
+     *
+     * @param      $defaultName
+     * @param null $gateway Optional, will default to current gateway if instantiated
+     *
+     * @return bool|string Returns false if no custom gateway field name has been defined, otherwise returns the custom
+     *                     name
+     */
+    public function getGatewayFieldName($defaultName, $gateway = null)
+    {
+        if (!$gateway) {
+            if (!$this->gateway) {
+                return false;
+            }
+
+            return $this->getGatewayFieldName($defaultName, $this->gateway);
         }
 
-        foreach($stack as $map) {
-            foreach ($map as $otherName => $defaultName) {
-                if ($hasPrefix) {
-                    $otherName = $renameMap['prefix'] . $otherName;
-                }
-                if (array_key_exists($otherName, $data)) {
-                    $value = $data[$otherName];
-                    unset($data[$otherName]);
-                    $data[$defaultName] = $value;
-                }
+        $renameMap = $this->getRenameConfig();
+
+        if (is_array($renameMap) && array_key_exists($gateway, $renameMap)) {
+            $gatewayMap = $renameMap[ $gateway ];
+            if (array_key_exists($defaultName, $gatewayMap)) {
+                return $gatewayMap[ $defaultName ];
             }
         }
 
-        return $data;
+        return false;
+    }
+
+    /**
+     * DRY helper to fetch the rename configuration
+     * @return array|bool|\scalar
+     */
+    public function getRenameConfig()
+    {
+        $renameMap = \Config::inst()->get('GatewayFieldsFactory', 'rename');
+
+        if (!$renameMap || empty($renameMap)) {
+            return false;
+        }
+
+        return $renameMap;
+
+    }
+
+    /**
+     * Builds the rename map which is used as a lookup table for normalizeFieldData()
+     * @return void
+     */
+    public function buildRenameMap() {
```

Comment:
I think this should have it's visibility changed to private or protected.
And I also think that this method should do the heavy lifting, eg. contain all the logic in `getFieldName`, so that `getFieldName` itself can be just a simple lookup. Reasoning: `getFieldName` is being called for every field that is being generated, whereas `buildRenameMap` will (usually) only be called once.